File: app/src/main/java/com/kelsos/mbrc/ui/activities/BaseActivity.java
package com.kelsos.mbrc.ui.activities;

import android.app.ActivityManager;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.design.widget.Snackbar;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.TaskStackBuilder;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.google.inject.Inject;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.constants.UserInputEventType;
import com.kelsos.mbrc.controller.RemoteService;
import com.kelsos.mbrc.events.MessageEvent;
import com.kelsos.mbrc.events.ui.ConnectionStatusChange;
import com.kelsos.mbrc.events.ui.DisplayDialog;
import com.kelsos.mbrc.events.ui.NotifyUser;
import com.kelsos.mbrc.ui.activities.nav.LibraryActivity;
import com.kelsos.mbrc.ui.activities.nav.LyricsActivity;
import com.kelsos.mbrc.ui.activities.nav.MainActivity;
import com.kelsos.mbrc.ui.activities.nav.NowPlayingActivity;
import com.kelsos.mbrc.ui.activities.nav.PlaylistActivity;
import com.kelsos.mbrc.ui.dialogs.SetupDialogFragment;
import com.kelsos.mbrc.ui.dialogs.UpgradeDialogFragment;
import com.squareup.otto.Bus;
import com.squareup.otto.Subscribe;
import roboguice.RoboGuice;

public abstract class BaseActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {
  @Inject Bus bus;

  @BindView(R.id.toolbar) Toolbar toolbar;
  @BindView(R.id.drawer_layout) DrawerLayout drawer;
  @BindView(R.id.nav_view) NavigationView navigationView;

  private TextView connectText;
  private ActionBarDrawerToggle toggle;
  private DialogFragment mDialog;

  private boolean isMyServiceRunning(Class<?> serviceClass) {
    ActivityManager manager = (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);
    for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
      if (serviceClass.getName().equals(service.service.getClassName())) {
        return true;
      }
    }
    return false;
  }

  protected abstract int active();

  private boolean onConnectLongClick(View view) {
    ifNotRunningStartService();
    bus.post(new MessageEvent(UserInputEventType.ResetConnection));
    return true;
  }

  private void onConnectClick(View view) {
    ifNotRunningStartService();
    bus.post(new MessageEvent(UserInputEventType.StartConnection));
  }

  private void ifNotRunningStartService() {
    if (!isMyServiceRunning(RemoteService.class)) {
      startService(new Intent(this, RemoteService.class));
    }
  }

  @Override protected void onDestroy() {
    super.onDestroy();
    drawer.removeDrawerListener(toggle);
    RoboGuice.destroyInjector(this);
  }

  @Override public void onBackPressed() {
    if (drawer.isDrawerOpen(GravityCompat.START)) {
      drawer.closeDrawer(GravityCompat.START);
    } else {
      super.onBackPressed();
    }
  }

  @Override public void onConfigurationChanged(Configuration newConfig) {
    super.onConfigurationChanged(newConfig);
    toggle.onConfigurationChanged(newConfig);
  }

  @Override protected void onPostCreate(Bundle savedInstanceState) {
    super.onPostCreate(savedInstanceState);
    toggle.syncState();
  }

  @Override public boolean onKeyUp(int keyCode, @NonNull KeyEvent event) {
    switch (keyCode) {
      case KeyEvent.KEYCODE_VOLUME_UP:
        return true;
      case KeyEvent.KEYCODE_VOLUME_DOWN:
        return true;
      default:
        return super.onKeyUp(keyCode, event);
    }
  }

  @Subscribe public void onConnection(ConnectionStatusChange event) {
    int resId;
    switch (event.getStatus()) {
      case CONNECTION_OFF:
        resId = R.string.drawer_connection_status_off;
        break;
      case CONNECTION_ON:
        resId = R.string.drawer_connection_status_on;
        break;
      case CONNECTION_ACTIVE:
        resId = R.string.drawer_connection_status_active;
        break;
      default:
        resId = R.string.drawer_connection_status_off;
        break;
    }

    connectText.setText(resId);
  }

  @Subscribe public void showSetupDialog(DisplayDialog event) {
    if (mDialog != null) {
      return;
    }
    if (event.getDialogType() == DisplayDialog.SETUP) {
      mDialog = new SetupDialogFragment();
      mDialog.show(getSupportFragmentManager(), "SetupDialogFragment");
    } else if (event.getDialogType() == DisplayDialog.UPGRADE) {
      mDialog = new UpgradeDialogFragment();
      mDialog.show(getSupportFragmentManager(), "UpgradeDialogFragment");
    } else if (event.getDialogType() == DisplayDialog.INSTALL) {
      mDialog = new UpgradeDialogFragment();
      ((UpgradeDialogFragment) mDialog).setNewInstall(true);
      mDialog.show(getSupportFragmentManager(), "UpgradeDialogFragment");
    }
  }

  @Subscribe public void handleUserNotification(NotifyUser event) {
    final String message = event.isFromResource() ? getString(event.getResId()) : event.getMessage();

    View focus = getCurrentFocus();
    if (focus != null) {
      Snackbar.make(focus, message, Snackbar.LENGTH_SHORT).show();
    }
  }

  @Override public boolean onKeyDown(int keyCode, KeyEvent event) {
    switch (keyCode) {
      case KeyEvent.KEYCODE_VOLUME_UP:
        bus.post(new MessageEvent(UserInputEventType.KeyVolumeUp));
        return true;
      case KeyEvent.KEYCODE_VOLUME_DOWN:
        bus.post(new MessageEvent(UserInputEventType.KeyVolumeDown));
        return true;
      default:
        return super.onKeyDown(keyCode, event);
    }
  }

  @Override public boolean onNavigationItemSelected(MenuItem item) {
    int itemId = item.getItemId();
    drawer.closeDrawer(GravityCompat.START);
    navigate(itemId);
    return true;
  }

  private void navigate(int itemId) {

    if (active() == itemId) {
      return;
    }

    if (itemId == R.id.nav_home) {
      createBackStack(new Intent(this, MainActivity.class));
    } else if (itemId == R.id.nav_library) {
      createBackStack(new Intent(this, LibraryActivity.class));
    } else if (itemId == R.id.nav_now_playing) {
      createBackStack(new Intent(this, NowPlayingActivity.class));
    } else if (itemId == R.id.nav_playlists){
      createBackStack(new Intent(this, PlaylistActivity.class));
    } else if (itemId == R.id.nav_lyrics) {
      createBackStack(new Intent(this, LyricsActivity.class));
    } else if (itemId == R.id.nav_settings) {
      startActivity(new Intent(this, SettingsActivity.class));
    } else if (itemId == R.id.nav_help) {
      Intent openHelp = new Intent(Intent.ACTION_VIEW);
      openHelp.setData(Uri.parse("http://kelsos.net/musicbeeremote/help/"));
      startActivity(openHelp);
    } else if (itemId == R.id.nav_feedback) {
      startActivity(new Intent(this, FeedbackActivity.class));
    } else if (itemId == R.id.nav_exit) {
      stopService(new Intent(this, RemoteService.class));
      finish();
    }
  }

  private void createBackStack(Intent intent) {
    TaskStackBuilder builder = TaskStackBuilder.create(this);
    builder.addNextIntentWithParentStack(intent);
    builder.startActivities();
  }

  /**
   * Should be called after RoboGuice injections and Butterknife bindings.
   */
  public void setup() {
    ifNotRunningStartService();
    setSupportActionBar(toolbar);

    toggle = new ActionBarDrawerToggle(this, drawer, toolbar, R.string.drawer_open, R.string.drawer_close);
    drawer.addDrawerListener(toggle);
    drawer.setDrawerShadow(R.drawable.drawer_shadow, GravityCompat.START);
    toggle.syncState();
    navigationView.setNavigationItemSelectedListener(this);

    View header = navigationView.getHeaderView(0);
    connectText = ButterKnife.findById(header, R.id.nav_connect_text);

    LinearLayout navConnect = ButterKnife.findById(header, R.id.nav_connect);
    navConnect.setOnClickListener(this::onConnectClick);
    navConnect.setOnLongClickListener(this::onConnectLongClick);

    ActionBar actionBar = getSupportActionBar();
    if (actionBar != null) {
      actionBar.setDisplayHomeAsUpEnabled(true);
      actionBar.setHomeButtonEnabled(true);
    }

    navigationView.setCheckedItem(active());
  }
}

File: app/src/main/java/com/kelsos/mbrc/BaseActivity.kt
package com.kelsos.mbrc

import android.content.Intent
import android.content.Intent.FLAG_ACTIVITY_CLEAR_TOP
import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.view.KeyEvent
import android.view.MenuItem
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import androidx.annotation.ColorRes
import androidx.annotation.LayoutRes
import androidx.annotation.StringRes
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.core.app.TaskStackBuilder
import androidx.core.content.ContextCompat
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.navigation.NavigationView
import com.google.android.material.snackbar.Snackbar
import com.kelsos.mbrc.common.state.ConnectionStateFlow
import com.kelsos.mbrc.common.state.ConnectionStatus
import com.kelsos.mbrc.features.help.HelpFeedbackActivity
import com.kelsos.mbrc.features.library.LibraryActivity
import com.kelsos.mbrc.features.lyrics.LyricsActivity
import com.kelsos.mbrc.features.nowplaying.NowPlayingActivity
import com.kelsos.mbrc.features.output.OutputSelectionDialog
import com.kelsos.mbrc.features.player.PlayerActivity
import com.kelsos.mbrc.features.playlists.PlaylistActivity
import com.kelsos.mbrc.features.radio.RadioActivity
import com.kelsos.mbrc.features.settings.SettingsActivity
import com.kelsos.mbrc.networking.ClientConnectionUseCase
import com.kelsos.mbrc.networking.client.UiMessage
import com.kelsos.mbrc.networking.client.UiMessageQueue
import com.kelsos.mbrc.networking.protocol.VolumeModifyUseCase
import com.kelsos.mbrc.platform.RemoteService
import com.kelsos.mbrc.platform.ServiceChecker
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject
import org.koin.androidx.scope.ScopeActivity
import timber.log.Timber

private const val NAVIGATION_DELAY = 250L

abstract class BaseActivity(
  @LayoutRes val contentLayoutId: Int,
) : ScopeActivity(contentLayoutId),
  NavigationView.OnNavigationItemSelectedListener {
  private val serviceChecker: ServiceChecker by inject()
  private val connectionUseCase: ClientConnectionUseCase by inject()
  private val volumeModifyUseCase: VolumeModifyUseCase by inject()
  private val connectionState: ConnectionStateFlow by inject()
  private val uiMessageQueue: UiMessageQueue by inject()

  private lateinit var toolbar: MaterialToolbar
  private lateinit var drawer: DrawerLayout
  lateinit var navigationView: NavigationView

  private lateinit var connectText: TextView
  private lateinit var toggle: ActionBarDrawerToggle
  private lateinit var connect: ImageView

  protected abstract fun active(): Int

  private fun onConnectLongClick(): Boolean {
    Timber.v("Connect long pressed")
    serviceChecker.startServiceIfNotRunning()
    connectionUseCase.connect()
    return true
  }

  protected fun onConnectClick() {
    Timber.v("Attempting to connect")
    serviceChecker.startServiceIfNotRunning()
    connectionUseCase.connect()
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    setupBackButtonHandler()

    toolbar = findViewById(R.id.toolbar)
    drawer = findViewById(R.id.drawer_layout)
    navigationView = findViewById(R.id.nav_view)

    setSupportActionBar(toolbar)

    toggle =
      ActionBarDrawerToggle(this, drawer, toolbar, R.string.drawer_open, R.string.drawer_close)
    drawer.addDrawerListener(toggle)
    toggle.syncState()
    navigationView.setNavigationItemSelectedListener(this)

    val header = navigationView.getHeaderView(0)
    connectText = header.findViewById(R.id.nav_connect_text)
    connect = header.findViewById(R.id.connect_button)
    connect.setOnClickListener { this.onConnectClick() }
    connect.setOnLongClickListener { this.onConnectLongClick() }

    supportActionBar?.setDisplayHomeAsUpEnabled(true)
    supportActionBar?.setHomeButtonEnabled(true)
    navigationView.setCheckedItem(active())
    serviceChecker.startServiceIfNotRunning()

    observeFlows()
  }

  override fun onDestroy() {
    super.onDestroy()
    drawer.removeDrawerListener(toggle)
  }

  override fun onConfigurationChanged(newConfig: Configuration) {
    super.onConfigurationChanged(newConfig)
    toggle.onConfigurationChanged(newConfig)
  }

  override fun onPostCreate(savedInstanceState: Bundle?) {
    super.onPostCreate(savedInstanceState)
    toggle.syncState()
  }

  override fun onKeyUp(
    keyCode: Int,
    event: KeyEvent,
  ): Boolean =
    when (keyCode) {
      KeyEvent.KEYCODE_VOLUME_UP -> true
      KeyEvent.KEYCODE_VOLUME_DOWN -> true
      else -> super.onKeyUp(keyCode, event)
    }

  private fun updateConnectionState(event: ConnectionStatus) {
    Timber.v("Handling new connection status %s", event.status)
    @StringRes val resId: Int

    @ColorRes val colorId: Int
    when (event) {
      ConnectionStatus.Offline -> {
        resId = R.string.drawer_connection_status_off
        colorId = R.color.black
      }

      ConnectionStatus.Authenticating -> {
        resId = R.string.drawer_connection_status_on
        colorId = R.color.accent
      }

      ConnectionStatus.Connected -> {
        resId = R.string.drawer_connection_status_active
        colorId = R.color.power_on
      }
    }

    connectText.setText(resId)
    connect.setColorFilter(ContextCompat.getColor(this, colorId))
  }

  override fun onKeyDown(
    keyCode: Int,
    event: KeyEvent,
  ): Boolean {
    val result =
      when (keyCode) {
        KeyEvent.KEYCODE_VOLUME_UP -> {
          lifecycleScope.launch {
            volumeModifyUseCase.increase()
          }
          true
        }

        KeyEvent.KEYCODE_VOLUME_DOWN -> {
          lifecycleScope.launch {
            volumeModifyUseCase.decrease()
          }
          true
        }

        else -> super.onKeyDown(keyCode, event)
      }
    return result
  }

  override fun onNavigationItemSelected(item: MenuItem): Boolean {
    val itemId = item.itemId
    drawer.closeDrawer(GravityCompat.START)
    drawer.postDelayed({ navigate(itemId) }, NAVIGATION_DELAY)
    return true
  }

  internal fun navigate(itemId: Int) {
    if (active() == itemId) {
      return
    }

    when (itemId) {
      R.id.nav_home -> {
        val upIntent = getParentActivityIntent()
        if (upIntent == null || shouldUpRecreateTask(upIntent)) {
          createBackStack(Intent(this, PlayerActivity::class.java))
        } else {
          navigateUpTo(upIntent)
        }
      }

      R.id.nav_library -> {
        createBackStack(Intent(this, LibraryActivity::class.java))
      }

      R.id.nav_now_playing -> {
        createBackStack(Intent(this, NowPlayingActivity::class.java))
      }

      R.id.nav_playlists -> {
        createBackStack(Intent(this, PlaylistActivity::class.java))
      }

      R.id.nav_radio -> {
        createBackStack(Intent(this, RadioActivity::class.java))
      }

      R.id.nav_lyrics -> {
        createBackStack(Intent(this, LyricsActivity::class.java))
      }

      R.id.nav_settings -> {
        createBackStack(Intent(this, SettingsActivity::class.java))
      }

      R.id.nav_help -> {
        createBackStack(Intent(this, HelpFeedbackActivity::class.java))
      }

      R.id.nav_output -> {
        OutputSelectionDialog.create(supportFragmentManager).show()
      }

      R.id.nav_exit -> {
        exitApplication()
      }
    }
  }

  internal fun exitApplication() {
    if (!RemoteService.serviceStopping) {
      val intent = Intent(this, RemoteService::class.java)
      stopService(intent)
    }

    if (this is PlayerActivity) {
      finish()
    } else {
      val intent = Intent(this, PlayerActivity::class.java)
      intent.putExtra(EXIT_APP, true)
      intent.flags = FLAG_ACTIVITY_CLEAR_TOP
      startActivity(intent)
    }
  }

  private fun createBackStack(intent: Intent) {
    val builder = TaskStackBuilder.create(this)
    builder.addNextIntentWithParentStack(intent)
    builder.startActivities()
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
      overrideActivityTransition(OVERRIDE_TRANSITION_OPEN, R.anim.slide_in, R.anim.slide_out)
    } else {
      @Suppress("DEPRECATION")
      overridePendingTransition(R.anim.slide_in, R.anim.slide_out)
    }
  }

  private fun observeFlows() {
    lifecycleScope.launch {
      repeatOnLifecycle(Lifecycle.State.STARTED) {
        connectionState.connection.collect { event ->
          updateConnectionState(event)
        }
      }
    }

    lifecycleScope.launch {
      repeatOnLifecycle(Lifecycle.State.STARTED) {
        uiMessageQueue.messages.collect {
          if (it is UiMessage.PluginUpdateRequired) {
            showPluginUpdateRequired(it.minimumVersion)
          } else if (it is UiMessage.PluginUpdateAvailable) {
            showPluginUpdateAvailable()
          }
        }
      }
    }
  }

  fun showPluginUpdateRequired(minimumRequired: String) {
    val intent = Intent(this, UpdateRequiredActivity::class.java)
    intent.putExtra(UpdateRequiredActivity.VERSION, minimumRequired)
    startActivity(intent)
  }

  fun showPluginUpdateAvailable() {
    val snackBar =
      Snackbar.make(
        navigationView,
        R.string.main__dialog_plugin_outdated_message,
        Snackbar.LENGTH_INDEFINITE,
      )
    snackBar.setAction(android.R.string.ok) { snackBar.dismiss() }
    snackBar.show()
  }

  private fun setupBackButtonHandler() {
    onBackPressedDispatcher.addCallback(
      this,
      object : OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
          if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
          } else {
            finish()
          }
        }
      },
    )
  }

  companion object {
    const val EXIT_APP = "mbrc.exit"
  }
}
